auth/src/main/java/com/firebase/ui/auth/viewmodel/email/WelcomeBackPasswordHandler.java

```diff
@@ -0,0 +1,183 @@
+    }
+
+    public boolean onActivityResult(int requestCode, int resultCode, Intent data) {
+        if (requestCode == RC_SAVE) {
+            if (resultCode != Activity.RESULT_OK) {
+                Log.e(TAG, "SAVE: Canceled by user");
+            }
+
+            mPendingResolutionLiveData.setValue(null);
+            setSuccess(mPendingIdpResponse);
+
+            return true;
+        }
+
+        return false;
+    }
+
+    public void startSignIn(@NonNull final String email,
+                            @NonNull final String password,
+                            @NonNull final IdpResponse inputResponse,
+                            @Nullable final AuthCredential credential) {
+        mSignInLiveData.setValue(new Resource<IdpResponse>());
+
+        // Build appropriate IDP response based on inputs
+        final IdpResponse outputResponse;
+        if (credential == null) {
+            outputResponse = new IdpResponse.Builder(
+                    new User.Builder(EmailAuthProvider.PROVIDER_ID, email).build())
+                    .build();
+        } else {
+            outputResponse = new IdpResponse.Builder(inputResponse.getUser())
+                    .setToken(inputResponse.getIdpToken())
+                    .setSecret(inputResponse.getIdpSecret())
+                    .build();
+        }
+
+        // Kick off the flow including signing in, linking accounts, and saving with SmartLock
+        getAuth().signInWithEmailAndPassword(email, password)
+                .addOnFailureListener(new TaskFailureLogger(TAG, "signInWithEmailAndPassword failed."))
+                .continueWithTask(new Continuation<AuthResult, Task<AuthResult>>() {
+                    @Override
+                    public Task<AuthResult> then(@NonNull Task<AuthResult> task) throws Exception {
+                        // Forward task failure
+                        if (!task.isSuccessful()) {
+                            throw task.getException();
+                        }
+
+                        // Task succeeded, link user if necessary
+                        if (credential == null) {
+                            return Tasks.forResult(task.getResult());
+                        } else {
+                            return task.getResult().getUser()
+                                    .linkWithCredential(credential)
+                                    .continueWithTask(new ProfileMerger(outputResponse))
+                                    .addOnFailureListener(new TaskFailureLogger(TAG, "linkWithCredential+merge failed."));
+                        }
+                    }
+                })
+                .addOnCompleteListener(new OnCompleteListener<AuthResult>() {
+                    @Override
+                    public void onComplete(@NonNull Task<AuthResult> task) {
+                        if (!task.isSuccessful()) {
+                            mSignInLiveData.setValue(new Resource<IdpResponse>(task.getException()));
```

Comment:
It seems (from reading/ripping up SmartLockBase) we basically consider SmartLock operations to be best effort.